Add app suite use case (#3391)
FusionAuth/fusionauth-site: astro/src/content/docs/get-started/use-cases/app-suite.mdx

import AppSuiteDescription from 'src/content/docs/get-started/use-cases/_app-suite-description.mdx';
import Breadcrumb from 'src/components/Breadcrumb.astro';

## Overview

<AppSuiteDescription />

## Problem 

You have multiple applications that you want to share across a common user base.

These applications may:

* be hosted at different domains or under different paths
* use SAML or OIDC for authentication
* be custom, commercial or open source
* be written using different technology stacks

You want to let your users log in once, and switch between applications without having to log in again. Users have different roles for each application.

## Solution

With the app suite use case, after the user logs in to one application, they are logged in to all of them.

You'll also have one customer profile and can view their login activity across the applications.

## Prerequisites

You have configured your application or applications to [delegate authentication to FusionAuth](/docs/get-started/use-cases/auth-service) via the Authorization Code grant and the hosted login pages.

## Example Scenario

Suppose you have a thriving e-commerce business with multiple applications:

> 

Author: Opened a request for this image.

* an ecommerce store where people can buy clown costumes
* a virtual try-on app which lets you see yourself in the outfits before you buy
* a forum for discussion about the latest trends
* a ticketing system for customer support

All of these use FusionAuth for their login. Each is represented by a different application, but your users have to log in each time they switch between applications.

## Actors/Components

* your user and their client application (mobile app or browser)
* two or more applications
* FusionAuth

This use case is built on cookies and redirects to FusionAuth. The redirect is typically transparent to the end user.

## Implementation Steps

This is a two step implementation process to ensure customers can access all the applications. Steps to take:

> * Ensure that end users have their FusionAuth SSO session enabled when they first authenticate.

Author: I think I get to that just below, so going to omit this.

> * Have applications to forward users to FusionAuth when they are not logged in to the application.

Author: same thing, I get to that below. Want to keep these step outlines simple.

### Force SSO Session Creation

The easiest way to force the SSO session to be created for each user authentication is to modify the login page of the theme. Rather than allow the user to check or uncheck the `rememberDevice` checkbox, which is the default setting, update the theme to always have that set to the value of `true`.

> To do so, update the `OAuth Authorize` template. This is the login page. 

Author: good call, added.

Here's a sample FreeMarker block to replace the default `rememberDevice` checkbox.

```
[@hidden name="rememberDevice" value="true"/]
```

The duration of this SSO session is controlled in the tenant settings.

### Forward Unauthenticated Users

Whenever a user accesses an application and is not logged in, forward them to the login URL for FusionAuth.

> Since your applications are already delegating authentication to FusionAuth, use the same logic to create the login URL.

Author: Updated to point to /docs/lifecycle/authenticate-users/oauth/endpoints#authorization-code-grant-request which I think is a bit more applicable. Good call.

Here's a diagram of a user switching between the clownwear ecommerce store and the forum.

<ApplicationSwitchDiagram alt="An example of an application switch." />

If a user is browsing the ecommerce store, then clicks on the forum link, they will transparently be logged in to the forum application.

Their client will get a token associated with the forum application, including the roles the user has for that application and claims such as the audience claim (`aud`) set correctly.

## Expected Outcome

Users switch between your applications transparently and securely. 

## Edge Cases

> SSO login works across all apps in a tenant, including those the user isn't registered for. After bouncing to FusionAuth, the user will be authenticated, but the `applicationId` and `roles` claims will be missing from the access token. Your applications should check these claims when determining whether a user is authorized for an application. [Read more about authentication and authorization](/docs/get-started/core-concepts/authentication-authorization).

Author: I'm not sure I understand how adding the self-service registration note is helpful. It doesn't matter how the user gets registered for an application, the point of this limitation is that the application needs to check claims to ensure the user is registered.

What am I missing?

When your user logs out, destroy the SSO session as each application session. You can do that using API calls or with FusionAuth's Front-Channel logout. [Read more about logout](/docs/lifecycle/authenticate-users/logout-session-management).

Any applications that have self-service registration enabled will automatically register users for with default roles when the user is bounced to that application. 

If an application only wants to allow logins from a certain identity provider or impose other conditions, you'll need extra integration work. For example, imagine there's a time tracking app for your clownwear ecommerce store that employees log in to using Google Workspace federation. You don't want customers to be able to view this application. One way you can prevent this with a [login validation lambda](/docs/extend/code/lambdas/login-validation).

> This use case only works with FusionAuth hosted login pages. If you use the Login API, you need to build your SSO. See this [GitHub issue for more details](https://github.com/fusionauth/fusionauth-issues/issues/1515).

Reviewer: ```suggestion
This use case only works with FusionAuth hosted login pages. If you use the Login API, you need to build your SSO. See this [GitHub issue for more details](https://github.com/fusionauth/fusionauth-issues/issues/1515).
```

I am not sure why, but I don't love linking this complex issue here.  This SSO creation is complex and I don't think that this GH issue explains the problem space fully.  We might solve SSO a different way / integrate existing SSO with the APIs in a different way in the future.  Dealers choice I suppose 😄

Author: Fair enough. Removed the link.

> If one application, such as the ecommerce app, requires authentication checks such as multi-factor authentication, additional self-serve required profile attributes, or registration verification, and others do not, users will be prompted when visiting the ecommerce application. 

Reviewer: This sentence reads long

## Other Example Scenarios

Any set of applications which share customer or user accounts are a good fit for this use case.

> 

Reviewer: For example, .....

---
title: App Suite
description: Let your users log in once and then transparently be logged into others.
section: get started
subcategory: use cases
navOrder: 20
---
import ApplicationSwitchDiagram from 'src/diagrams/docs/get-started/use-cases/app-suite/application-switch.astro';
import AppSuiteDescription from 'src/content/docs/get-started/use-cases/_app-suite-description.mdx';
import Breadcrumb from 'src/components/Breadcrumb.astro';
import InlineField from 'src/components/InlineField.astro';

## Overview

<AppSuiteDescription />

## Problem 

> You have multiple applications that you want to share across a common user base.

Reviewer: ```suggestion
You have multiple applications you want to share across a common user base.
```

These applications may:

* be hosted at different domains or under different paths
* use SAML or OIDC for authentication
* be custom, commercial, or open source
* be written using different technology stacks

> You want to let your users log in once, and switch between applications without having to log in again. Users have different roles for each application.

Reviewer: ```suggestion
You want to let your users log in once, and switch between applications without having to log in again. Users may or may not have different roles for each application.
```

## Solution

> With the app suite use case, after the user logs in to one application, they are logged in to all of them.

Reviewer: ```suggestion
With the app suite use case, after the user logs in to one application, they are functionally logged in to every application via the magic sauce of FusionAuth SSO.
```

> With the app suite use case, after the user logs in to one application, they will transparently log in to every application via the magic sauce of FusionAuth SSO. You'll also have only one customer profile to manage. You can view their login activity across the applications.

Reviewer: ```suggestion
With the app suite use case, after the user logs in to one application, they will transparently log in to every application via the magic sauce of FusionAuth SSO. You'll also have only one customer profile to manage (this is a huge administrative win). You can view their login activity across the applications.
```

Reviewer: or : 

With the app suite use case, after the user logs in to one application, they will transparently log in to every application via the magic sauce of FusionAuth SSO. As a benefit, you'll also have only one customer profile to manage and you can view their login activity across the applications.

You'll also have one customer profile and can view their login activity across the applications.

## Prerequisites

> You have configured your application or applications to [delegate authentication to FusionAuth](/docs/get-started/use-cases/auth-service) via the Authorization Code grant and the hosted login pages.

Reviewer: ```suggestion
You have configured your application or applications to [delegate authentication to FusionAuth](/docs/get-started/use-cases/auth-service) via the Authorization Code grant and the hosted login pages.  In other words you are using OAuth2, our built in workflows, and JWT's to authenticate a user.
```

Author: I feel like this is maybe more detail that is needed, given the link.

> You have configured your application or applications to [delegate authentication to FusionAuth](/docs/get-started/use-cases/auth-service) via the Authorization Code grant and the FusionAuth hosted login pages.

Reviewer: ```suggestion
You have configured your application(s) to [delegate authentication to FusionAuth](/docs/get-started/use-cases/auth-service) via the Authorization Code grant and the FusionAuth hosted login pages.
```

> Suppose you have a thriving e-commerce business with multiple applications:

Reviewer: ```suggestion
Suppose you have a thriving e-commerce business with much clown love and with multiple end user applications:
```

Suppose you have a thriving e-commerce business with multiple applications:

* an ecommerce store where people can buy clown costumes
* a virtual try-on app which lets you see yourself in the outfits before you buy
* a forum for discussion about the latest trends
* a ticketing system for customer support

All of these use FusionAuth for their login. Each is represented by a different application and client Id, but your users have to log in each time they switch between applications.

## Actors/Components

* your user and their client application (mobile app or browser)
* two or more applications
* FusionAuth

This use case is built on cookies and redirects to FusionAuth. The redirect is typically transparent to the end user.

## Implementation Steps

This is a two step implementation process to ensure customers can access all the applications. Steps to take:

> * Ensure end users have their FusionAuth SSO session enabled when they first authenticate.

Author: I think this is duplicative :) . We're going to tell them how to implement this shortly.

> This use case is built on cookies and redirects to FusionAuth. The redirect is typically transparent to the end user.

Reviewer: ```suggestion
This use case is built on cookies and redirects to FusionAuth. The redirect is typically transparent to the end user (read: they won't notice it, but will appreciate the seamless integration).
```

> * Have applications forward users to FusionAuth when users are not logged in to the application.

Author: good point. Let me rework.

> This is a two step implementation process to ensure customers can access all the applications. Steps to take:

Reviewer: ```suggestion
This is a two step implementation process to ensure customers can access all the applications. Don't worry -- it's a quick process as there are only two steps to take:
```

* Ensure end users have their FusionAuth SSO session enabled when they first authenticate.
* When a user is not logged in to a particular application, forward the user to FusionAuth. They'll be transparently logged in and delivered back to the application.

### Force SSO Session Creation

The easiest way to force the SSO session to be created as a user logs in is to modify the login page. You do so by modifying the theme. In the default theme, a user can check or uncheck the `rememberDevice` checkbox, labelled by default "Keep me signed in". To force creation of the SSO session, update the theme to set the `rememberDevice` parameter to `true`. Modify the `OAuth Authorize` template [using the advanced theme editor](/docs/customize/look-and-feel/advanced-theme-editor)

Here's a sample FreeMarker block you can use to update the default template.

```
[@hidden name="rememberDevice" value="true"/]
```

> 

Reviewer: Do we need a note about Simple themes Folx being SOL for the time being?

Author: I don't think so. We talk about the advanced theme editor above.

> The duration of this SSO session is controlled in the tenant settings. You can modify it by navigating to <Breadcrumb>Tenant -> Your Tenant -> OAuth</Breadcrumb> and updating the <InlineField>Session timeout</InlineField>.

Reviewer: ```suggestion
The duration of this SSO session is controlled in the Tenant settings. You can modify it by navigating to <Breadcrumb>Tenant -> Your Tenant -> OAuth</Breadcrumb> and updating the <InlineField>Session timeout</InlineField>.
```

### Forward Unauthenticated Users

> Whenever a user is not logged in and accesses an application in your app suite, forward them to the login URL for FusionAuth. As your applications are already delegating authentication to FusionAuth, use the same logic or library functions to create the login URL. Here's an [explanation of how to create the URL and what each component means](/docs/lifecycle/authenticate-users/oauth/endpoints#authorization-code-grant-request).

Reviewer: ```suggestion
Whenever a user is not logged in and accesses an application in your app suite, forward them to the login URL for FusionAuth. As your applications are already delegating authentication to FusionAuth (see prerequisites above), use the same logic or library functions to create the login URL to forward the user onward. Here's an [explanation of how to create the URL and what each component means](/docs/lifecycle/authenticate-users/oauth/endpoints#authorization-code-grant-request).
```

Here's a diagram of a user switching between the clownwear ecommerce store and the forum.

<ApplicationSwitchDiagram alt="An example of an application switch." />

> If a user is browsing the ecommerce store, then clicks on the forum link, they will transparently be logged in to the forum application.

Reviewer: ```suggestion
If a user is browsing the e-commerce store, then clicks on the forum link, they will transparently be logged in to the forum application.
```

> Whenever a user is not logged in and accesses an application in your app suite, forward them to the login URL for FusionAuth. As your applications are already delegating authentication to FusionAuth, use the same logic or library functions to create it. Here's an [explanation of how to create the URL and what each component means](/docs/lifecycle/authenticate-users/oauth/endpoints#authorization-code-grant-request).

Reviewer: ```suggestion
Whenever a user is not logged in and accesses an application in your app suite, forward them to the login URL for FusionAuth. As your applications are already delegating authentication to FusionAuth, use the same logic or library functions to create an authorize URL. Here's an [explanation of how to create the URL and what each component means](/docs/lifecycle/authenticate-users/oauth/endpoints#authorization-code-grant-request).
```

> Their client will get a token associated with the forum application, including the roles the user has for that application and claims such as the audience claim (`aud`) set correctly. The user will be logged in to the new application, as desired.

Reviewer: ```suggestion
Their client will get a token associated with the forum application, including the roles the user has for that application and claims such as the audience claim (`aud`) set correctly. The user will be logged in to the new application, as desired (pretty cool, huh?).
```

## Expected Outcome

Users switch between your applications transparently and securely. 

## Edge Cases

SSO login works across all apps in a tenant, including those the user isn't registered for. After redirection to and from FusionAuth, the user will be authenticated, but the `applicationId` and `roles` claims will be missing from the access token. Applications must check these claims when determining whether a user is authorized for an application. [Read more about the difference between authentication and authorization](/docs/get-started/core-concepts/authentication-authorization).

When your user logs out, destroy the SSO session as well as each application session. You can do that using API calls or with FusionAuth's Front-Channel logout. [Read more about logout](/docs/lifecycle/authenticate-users/logout-session-management).

> Any applications that have self-service registration enabled will automatically register users for with default roles when the user is redirected to that application. 

Reviewer: ```suggestion
Any applications that have [self-service registration](https://fusionauth.io/docs/lifecycle/register-users/basic-registration-forms) enabled will automatically register users for with default roles when the user is redirected to that application. 
```

> If an application only wants to allow logins from a certain identity provider or impose other conditions, you'll need extra integration work. For example, imagine there's a time tracking app for your clownwear ecommerce store that employees log in to using Google Workspace federation. You don't want customers to be able to view this application. One way you can prevent this with a [login validation lambda](/docs/extend/code/lambdas/login-validation).

Reviewer: ```suggestion
If an application only wants to allow logins from a certain identity provider or impose other conditions, you'll need extra integration work. For example, imagine there's a time tracking app for your clownwear e-commerce store that employees log in to using Google Workspace federation. You don't want customers to be able to view this application. One way you can prevent this with a [login validation lambda](/docs/extend/code/lambdas/login-validation).
```

Reviewer: An other places with this ecommerce mention....

Author: Thanks. I will standardize on e-commerce.

The functionality described in this use case only works with [FusionAuth hosted login pages](/docs/get-started/core-concepts/hosted-login-vs-api-login#hosted-login-pages). If you use the Login API, you need to build your SSO implementation.

Applications may have different authentication related requirements. Examples include:

* multi-factor authentication (MFA)

> * required profile attributes

Reviewer: ```suggestion
* required registration fields (complete registration workflow)
```

> * registration verification

Reviewer: ```suggestion
* [registration verification](https://fusionauth.io/docs/lifecycle/manage-users/verification/registration-gate-accounts-until-verified)
```
